### PR TITLE
[FSDP] Remove unneeded stream sync from `clip_grad_norm_()`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#89308 [FSDP] Remove unneeded stream sync from `clip_grad_norm_()`**

We do not need to have the pre-unshard and unshard streams wait for the computation stream because we are not using the pre-unshard or unshard streams in `clip_grad_norm_()`.

The other change is simply avoiding a loop to get `grads`.

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1156,32 +1155,34 @@ def clip_grad_norm_(
...
             not_fsdp_managed = (
                 param not in sharded_params and param not in nonsharded_params
             )
             if not_fsdp_managed:
                 nonsharded_params.add(param)
+                if param.grad is not None:
+                    grads.append(param.grad)
         local_sharded_norm = _get_grad_norm(sharded_params, norm_type).to(
             self.compute_device
         )
```
Comment (mrshenli) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1188]: looks like the `_get_grad_norm` uses grad values. Where do we make sure that reduce scatter on grad is done?

**Review by mrshenli — APPROVED** [review]
stamp to unblock

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1188]: Post-backward final callback:
https://github.com/pytorch/pytorch/blob/c599cf24ad01a4befa58642f9e703792f53b79fa/torch/distributed/fsdp/_runtime_utils.py#L701-L702

**Pull request closed without merging**